### DISCUSSION TITLE
Standups - Sprint 2

### DISCUSSION THREAD

**Discussion by Justus-at-Tazama** [tazama-lf/General-Issues#75] [General]
## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

**@vorsterk - Kyle Vorster**

**@Lenbkr - Len Bekker**

**@JeanPierreNell - Jean-Pierre Nell**

**@kawaki-san - Rodney Kanjala**

**@johanfol - Johan Foley**

**Comment by Justus-at-Tazama** [comment]
## 2023/05/22

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@vorsterk - Kyle Vorster**

Yesterday: Working on a demo video for the EKS deployments. The Postman tests are failing and need to be updated for an AWS deployment. Working on Sandbox 2 to create a security token for APM. Struggling to get Logstash and APM connected to EKS. Connection resolved, but not logging on the dashboard now.

Today: Carrying on with performance benchmarking. Prep for demo (CI/CD pipeline).

**@Lenbkr - Len Bekker**

Yesterday: Wrote jdoc inline methods for the library methods. Aiming for reader-friendly. PR merged. Assisted JP with an issue on the deployment of a rule with a token issue. Helped Rodney on the implementation of a library query.

Today: Record demo video. Support for library troubleshooting.

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: The transactions completed eventually (after a few hours). Arango moved to its own node. APM deployed. Submitted 1000 transactions again. Much faster and now reduced to a crawl. APM logs went missing.

Today: Rerun the tests. Foley requested to run all the typologies. Record demo video for the PPA.

**@cshezi - Cebolenkosi Shezi**

Yesterday: Worked on Rule 007. Some questions on compiling the pre-scripts. Completed the first three scenarios. Recorded the demo video.

Today: Continuing with rule 007 and complete test scenarios. Finalise the preparation for the demo.

**@kawaki-san - Rodney Kanjala**

Yesterday: Worked on rule 020. Walked through the rule code. Addressed multiple issues. Will work through learnings in the retrospective.

Today: Adding benchmarking in Postman for the rule test scenarios. Compiling demo slides.

**@johanfol - Johan Foley**

Yesterday: Reports completed. CRSP updated. Completed presentation for demo.

Today: Demo video.

## Parking lot

1. Go through all the statuses of all the rules that everyone is working on and make sure it is up to date.
2. SonarCloud feedback on the Channel Router needs to be updated. Kyle please take a screenshot for the demo.

**Comment by Justus-at-Tazama** [comment]
## 2023/05/23

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

Yesterday: Rerecording demo video. Continued rule 007.

Today: Continuing with rule 007.

Blocker: Rule 007 is taking longer than intended - check if swarm required.

**@vorsterk - Kyle Vorster**

Yesterday: Completed and polished documentation. Working on getting logstash working. Writing to Kibana, but not showing in dashboards. Restarted Elastic masternode - won't restart due to a duplicate UUID issue.

Today: Supporting Beta test. Uninstall Kibana Helm chart to reset default values (ghosts).

Blocker: Need a new laptop

**@Lenbkr - Len Bekker**

Yesterday: Admin and maintenance. Please review rule 003's PR. Reviewed Foley's CRSP.

Today: Fix a bug in documentation. Looking into a bug that shortchanges the pulling in of the right package version of the library. Start with the splitting of the DB's ticket #71 .

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: Turned on the big network map for the beta test. Some issues with images that needed to be rebuilt. Split batches into 500. Only 11 documents passed on the full network map. 10 passed immediately and then the next one came through 45 minutes later. Repeated, same results. Kibana service down - can't see why yet. Some processors are unable to redeploy to utilise APM (002, CADProc). Foley requested to check for master nodes. A

Today: Carry on with caching while awaiting for the Kibana to be resolved.

Blockers: Kibana service down.

**@kawaki-san - Rodney Kanjala**

Yesterday: Integrate benchmarking into rule 020. Ran tests. Issues with dirty data in the development DB.

Today: Clean up the unit tests.

**@johanfol - Johan Foley**

Today: Prep for the code review - architecture sprint. Review upcoming roadmap.

## Parking lot

1. Redis issue may be related to clustering. Foley will attempt to replicate.
2. #71 check what is possible in open source. Timebox to research the problem and possible solutions for today.

**Comment by Justus-at-Tazama** [comment]
## 2023/05/22

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

Yesterday: Complete rule 007! Polishing and double-checking. Created a PR. Updated rule 008 as well. Len supported with a couple of issues. Resubmit PR with both rules.

Today: Starting rule 018 #38 

**@vorsterk - Kyle Vorster**

Yesterday: Setting up new laptop. The new laptop is great! But everything needs to be reinstalled. Logstash working in Sandbox. New Elastic is working much better.

Today: Benchmarks. Foley recommends to start small (low volume)

**@Lenbkr - Len Bekker**

Yesterday: Finalised sprint leftovers. Updated the library to implement appropriate semantic versioning. Investigated Arango - only available options are RocksDB tweaking and sharding. The free version limits our options. We can split the databases by collections, but there are limits to the benefits we will achieve this way. Consider a synthetic benchmark that tests the database in the way that we use it. Improved data modelling has limited projected benefits.

Today: Moving on to indexing.

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: Issues with the logging on the Beta test. Kyle supporting. Logging appears to be resolved now. Heavy load on the CRSP and the TADProc. There are still 504's in the logs. Also worked on caching into rule 010. Using the library to implement the query.

Today: Complete caching in rule 010., then beta testing.

**@kawaki-san - Rodney Kanjala**

Yesterday: Rolled back optimisation changes in rule 020 to achieve a like-for-like implementation. Introduced a but - Foley agreed to implement the updated and optimised rule to avoid any further work. Spend some time with Shezi reviewing the first sprint learnings. Wrote documentation for the personal access token.

Today: Finish up unit tests for rule 020. Moving on to rule 025.

**@johanfol - Johan Foley**

Yesterday: Admin work on the project roadmap. Working on the PoC for NATS.

Today: NATS PoC

## Parking lot
1. Meeting to review the data pipeline, caching and indexing.

**Comment by Justus-at-Tazama** [comment]
## 2023/05/25

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

Yesterday: Cleaning rule 008 and rule 007. PR's created and requested JP to test. More cleanup out of the review. Started Rule 018. Problems with EndToEndId in the testing. Foley supported and problem resolved. Second scenario was coded incorrectly. Double-checking before wrapping up. Researching a tool to record findings.

Today: Commit rule 007 and rule 008, wrapping up testing on rule 018. Rule 021 next.

**@vorsterk - Kyle Vorster**

Yesterday: Performance benchmarking. Started slow and ramping up. APM and logstash is timing out, regardless of allocated resources. Desktop research on improving performance - the latest version may help. Helm Charts updated.

Today: Recover Elastic, then carry on with performance testing.

**@Lenbkr - Len Bekker**

Yesterday: Design session for optimization. Assisted Rodney in review. Query analyzer shows the projected cost of a query. Deploying Jenkins on local to start splitting the database. This did not go well. Deploy Arango directly and manually.

Today: Splitting the databases.

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: Design session for DB optimisation. Submitted 500 transactions for the full network map. Only 38 passed. Rule 054 was a problem and was temporarily removed. Some rules are running very slow. TADProc is running exceptionally slow. Reviewing APM. Assisted Shezi with postman testing and PR review.

Today: Beta testing.

**@kawaki-san - Rodney Kanjala**

Yesterday: Wrapping up rule 020. All test scenarios working. Postman tests overlap other tests and those tests run during the commit that causes a conflict. Ran the commit when the system was quiet and it worked. Refactored rule 025 - the library called on LIMITS and the rule did not.

Today: Wrapping up rule 025; moving on to rule 028.

**@johanfol - Johan Foley**

Yesterday: Design session on optimization. NATS PoC. Making good progress. Throughput down to 1ms.

Today: More NATS.

## Parking lot

**Comment by Justus-at-Tazama** [comment]
## 2023/05/29

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

Yesterday: Clearing up Postman issues. Created pull requests for rule 018 and rule 021.

Today: Rule 027, supporting PR reviews of submitted requests. Assist Rodney in Jenkins build issue.

**@vorsterk - Kyle Vorster**

Yesterday: Performance benchmarking. Scaling up, no errors and timeouts so far! Created a job to clean Jenkins disk space on the sandboxes. Started keeping track of costing. 132 TPS, along with APM and Logstash on Typology 028.

Today: Benchmarking!

**@Lenbkr - Len Bekker**

Yesterday: Spent some time with Kyle to assist in the Arango deployment. Running 9 separate instances of Arango DB across their separate namespaces. Testing and fine-tuning postman. Assisted the team with Jenkins deployment on a different environment.

Today: Carrying on with #71 - Foley says that all changes to support performance should be deployed only after we have am updated baseline benchmark.

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: Created some indexes on Friday and ran some tests on Sandbox 2. Bug in CADProc that was fixed. Challenging deployment.

Today: Carrying on with the beta testing. Also working on #76 

**@kawaki-san - Rodney Kanjala**

Yesterday: Set up an environment on Jenkins with Len's help. Working on passing the builds as the last step in the definition of done. Unsure of the "build" requirement. Setting up rule 017. Process generated storage errors.

Today: Rule 017, 005 and 020.

Blocker: Storage errors during build.

**@johanfol - Johan Foley**

Yesterday: Kermis, duck fishing.

Friday: NATS PoC. Fixed a bug in CADProc. Got JetStream working from CRSP to the Rules (Fanning out has a different setup to fanning in). Working with scaling up transparently. JetStream slightly slower than NATS (due to assurances in the delivery). Raised an issue on the NATS support to double-check our expectations.

Today: Check TADProc for the same issue. NATS: Rules to typologies.

## Parking lot

Build requirement: If you can run the job to deploy to an environment. The link to dev is currently broken because Dev is used for the performance testing. The Jenkins job builds and deploys the rule processor. Set it to Sandbox 4. Deploy the branch that hasn't been merged yet to verify that we *can* deploy it. Once the PR is done, you revert the change to re-implement the (new) main branch.

**Reply by johanfol** [reply]
Issue for confirming JetStream performance https://github.com/nats-io/nats.js/issues/573

**Comment by Justus-at-Tazama** [comment]
## 2023/05/30

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

Yesterday: Testing the rule deployment to Jenkins was a challenge. Still missing one of the variables (database certificate). Created PRs including the environment variable to explicit. (Rule 004, 002, 008). PR ready to review this morning. Worked with Rodney to make sure we are all using the same npr file. Len supported. Used the same file for all the rules.

Today: All other rules to have the environment variable included. Call for support if help required. Make sure all the tests are passing. All the rules are pointing to the CQA branch. Jobs to be moved to main branch today.

**@vorsterk - Kyle Vorster**

Yesterday: Running the benchmarks - no performance improvements. Rule 003 and rule 008 had issues with their library. It looks like there's a redis connection that we need to instantiate in our deployments. Adding redis connections to the Jenkins job in Dev. Scaled up the master node and the replica node for redis.

Today: Instantiate redis connection and re-run the tests.

**@Lenbkr - Len Bekker**

Yesterday: Worked on deployment charts for the multiple instances of Arango. Passed to Kyle for deployment optimization. Issues with the library undefined when deploying Jenkins. Note that if we want to test that our deployments are cloud-ready, we simulate the cloud behaviour via Docker. Remember that if things are working in your own Docker, ensure that the build is also working online before submitting the PR. Worked on Postman for the creation of the database. Still need to add indexes. Need a design wrap-up session to double-check indexing.

Today: Wrapping up the database split. Documentation, and instructions on debugging with Docker.

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: Emotional roller-coaster. Beta testing environment appears to have some scaling issues. Scaled up CADProc. Transactions passed. Some rules were failing. Cascading failures from the rules to Typology Processor, then CADProc, to TADProc. CADProc is reporting an issue, but the logs indicate that it is successful. Error goes to ELK, but the CADProc itself reports success. Debugging continues. Some libraries are out of date. Some integration issues.

Today: Debugging CADProc, carrying on with the Beta test.

**@kawaki-san - Rodney Kanjala**

Yesterday: Testing the Jenkins rules to make sure the builds were compiling and passing. Rules 011, 017, 025 all passed. Set up an npm.src file that was universal. File wasn't working on Shezi's files. Len cast a spell and fixed it. Set up Docker on local to test local builds and pre-empt any deployment issues. Working on Rule 028 - some tests weren't compiling - fixed syntax. 5 out of 15 tests were failing.

Today: Completing rule 028.

**@johanfol - Johan Foley**

Yesterday: NATS setup.

Today: Testing and benchmarking today. Demo this afternoon.

## Parking lot

Len: Indexing design wrap-up: Draw pictures in PlantUML. Catch up later.

**Comment by Justus-at-Tazama** [comment]
## 2023/05/31

## Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

## Apologies
None

## Minutes:

**@cshezi - Cebolenkosi Shezi**

Yesterday: Couldn't merge the PRs from the day before. Len assisting. Spend some more time getting to know Docker. Investigated consoling and connecting to Redis. Can we have a get() function in the library to fetch data to make sure that Redis is connected?

Today: Slam those PRs home!

**@vorsterk - Kyle Vorster**

Yesterday: Performance testing - generating a lot of errors. The errors are not logged. The TADProc may be creating issues. For no reason. There has been no impacting change that would account for the change. The CRSP is taking a very long time to respond. Foley recommends to debug from the back to the front (starting at TADProc and working towards the CRSP).

Today: Benchmarking, and working with JP to see if the issues are the same.

**@Lenbkr - Len Bekker**

Yesterday: Completed designs for the database splitting. Implemented in Postman. The list/watchlist collections are left, but can be added quickly. Added indexes to the most looked-at properties. Added inverted indexes on CreDtTm. Arango 3.11 improved processing on some indexes. Inverted indexes may be better in certain situations. Testing to confirm. Using collection variables in Postman to accommodate 9 different sets of database variables. Worked on the library to add/improve some error handling. Updating documentation on npm.rc for Jenkins.

Today: Review how Postman changes affect data preparation. Supporting libraries.

**@JeanPierreNell - Jean-Pierre Nell**

Yesterday: Getting sadder by the day. The rollercoaster is no longer fun. Deployed CADProc in the beta-testing. All scaling is now suspended. New issue. The pipeline completes very inconsistently. The databasemanager library is unexpectedly undefined in a number of calls. Brainstormed with Len about placing the config into the library. Documented the brainstorm.

Today: A luta continua

Blocker: swarm to resolve issues.

**@kawaki-san - Rodney Kanjala**

Yesterday: Working on setting up Docker containers. Struggling to pass the tests for rule 028. Rule 017 PR awaiting review.

Today: Wrap up rule 028.

**@johanfol - Johan Foley**

Yesterday: JetStream and NATS demo. Everyone was very excited.

Today: Starting with the limited PoC/Implementation for rules to typology implementation.

## Parking lot

**Comment by vorsterk** [comment]
2023/06/01
Attendees
@johanfol - Johan Foley
@Lenbkr - Len Bekker
@JeanPierreNell - Jean-Pierre Nell
@cshezi - Cebolenkosi Shezi
@vorsterk - Kyle Vorster
@kawaki-san - Rodney Kanjala

Apologies
None

Minutes:
@cshezi - Cebolenkosi Shezi

Yesterday : Merged four 4 rules to main and switched the branch from jenkins pipe line script to use main branch, and submitted one of the rule to QA for further end 2 end test investigation, Reviewed Rodney's PR for rule-025 and Worked on rule-018 and sent PR to Rodney for review

Today : Refactorizing rule-025 and after moving to rule-001

@vorsterk - Kyle Vorster

Yesterday: Performance testing  , Redis issues , Half day

Today: Deploying redis cluster and performance testing.

@Lenbkr - Len Bekker

Yesterday: Added isReadyCheck to Library. Helped JP. Figure out env and config for split DB logic.

Today: Update Lib to cater for new rule requirements. 

@JeanPierreNell - Jean-Pierre Nell

Yesterday: Fixed missing Rules, Optimised some rules with indexes. Started running batches, no missing transactions. Started on next ticket #76.

Today: Continue Ticket #76

Blocker: swarm to resolve issues.

@kawaki-san - Rodney Kanjala

Yesterday: Looking at dockerfile how the config params are populated,
Discovered that we use 'sed' and theres substitution involved. Adjusted Dockerfiles so that they match that pattern
Library refactor (dev quality of life changes)
Reviewed PR from Shezi, left some suggestions which I went over with him.
Refactored Rule 34, there were some TODO comments which I think the previous dev forgot to remove everything seemed to be implemented
Read up on the logic flow as a sanity check.

Today: Make sure there are no open handles in Rule34's tests (done). Cleaning up linting/formatting for rule35 and 36 as I await lib update so I can get rid of their local arango calls

@johanfol - Johan Foley

Yesterday: JetStream and NATS demo. Everyone was very excited.

Today: Starting with the limited PoC/Implementation for rules to typology implementation.

Parking lot
